cpp/ql/lib/experimental/semmle/code/cpp/dataflow/ProductFlow.qll

```diff
@@ -352,32 +354,63 @@ module ProductFlow {
         pragma[only_bind_out](succ.getNode().getEnclosingCallable())
     }
 
+    newtype TKind =
+      TInto(DataFlowCall call) {
+        [any(Flow1::PathNode n).getNode(), any(Flow2::PathNode n).getNode()]
+            .(ArgumentNode)
+            .getCall() = call
+      } or
+      TOutOf(DataFlowCall call) {
+        [any(Flow1::PathNode n).getNode(), any(Flow2::PathNode n).getNode()].(OutNode).getCall() =
+          call
+      }
+
     pragma[nomagic]
     private predicate interprocEdge1(
-      Declaration predDecl, Declaration succDecl, Flow1::PathNode pred1, Flow1::PathNode succ1
+      Declaration predDecl, Declaration succDecl, Flow1::PathNode pred1, Flow1::PathNode succ1,
+      TKind kind
     ) {
       Flow1::PathGraph::edges(pred1, succ1) and
       predDecl != succDecl and
       pred1.getNode().getEnclosingCallable() = predDecl and
-      succ1.getNode().getEnclosingCallable() = succDecl
+      succ1.getNode().getEnclosingCallable() = succDecl and
+      exists(DataFlowCall call |
+        kind = TInto(call) and
+        pred1.getNode().(ArgumentNode).getCall() = call and
+        succ1.getNode() instanceof ParameterNode
+        or
+        kind = TOutOf(call) and
+        succ1.getNode().(OutNode).getCall() = call and
+        pred1.getNode() instanceof ReturnNode
```

Comment:
There can be other out-flow besides the `OutNode`-to-`ReturnNode` pairs. It can also be from a `PostUpdateNode` that has flow from a parameter to a `PostUpdateNode` for an argument.

Reply:
Oh, yes. That's a good point. I decided to blatantly ignore jump steps for now, but the `PostUpdateNode` -> `PostUpdateNode` flow caused by setter-like functions is probably too frequent of an occurrence to be ignored.

Reply:
You can probably account for jump-steps by having a kind "jump" and identify that as an interproc edge that's not in one of the other cases.

Reply:
Good idea. Hopefully I've done this in 0d6fdc674bb92a8013ba889b34cf9c9e89180c0c 🤞.